# CALENDAR DATASET EXPLORATION USING PYTHON

# 📅 Calendar Data Validation

This notebook validates the calendar dataset used in the Coffee Shop BI project.

The goal is to ensure complete and consistent time coverage for accurate time-based analysis.

In [5]:
!pip install pandas
import os
import pandas as pd

Defaulting to user installation because normal site-packages is not writeable



[notice] A new release of pip is available: 26.0.1 -> 26.1
[notice] To update, run: C:\Users\keren mayer\AppData\Local\Microsoft\WindowsApps\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\python.exe -m pip install --upgrade pip


## 1. Data Loading

In [7]:

file_path = "OneDrive/Documents/Bureau/EXPERIS/FINAL BI PROJECT/data/us_calendar_jan_jun_2023_with_religious_holidays_time.xlsx"
df_calendar = pd.read_excel(file_path)

## 2. Initial Data Exploration

In [8]:
df_calendar.head()

,date,year,month,month_name,week_number,weekday_number,weekday_name,is_weekend,is_us_federal_holiday,holiday_name,is_religious_holiday,religious_holiday_name,religion,religious_is_estimated,time
0,2023-01-01,2023,1,January,52,7,Sunday,True,False,NaN,False,NaN,NaN,False,00:00:00
1,2023-01-01,2023,1,January,52,7,Sunday,True,False,NaN,False,NaN,NaN,False,01:00:00
2,2023-01-01,2023,1,January,52,7,Sunday,True,False,NaN,False,NaN,NaN,False,02:00:00
3,2023-01-01,2023,1,January,52,7,Sunday,True,False,NaN,False,NaN,NaN,False,03:00:00
4,2023-01-01,2023,1,January,52,7,Sunday,True,False,NaN,False,NaN,NaN,False,04:00:00


#### Inspecting the structure of the calendar dataset

In [9]:
df_calendar.shape

(4344, 15)

In [7]:
df_calendar.columns

Index(['date', 'year', 'month', 'month_name', 'week_number', 'weekday_number',
       'weekday_name', 'is_weekend', 'is_us_federal_holiday', 'holiday_name',
       'is_religious_holiday', 'religious_holiday_name', 'religion',
       'religious_is_estimated', 'time'],
      dtype='object')

## 3. Data Structure Analysis

In [8]:
df_calendar.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4344 entries, 0 to 4343
Data columns (total 15 columns):
 #   Column                  Non-Null Count  Dtype         
---  ------                  --------------  -----         
 0   date                    4344 non-null   datetime64[ns]
 1   year                    4344 non-null   int64         
 2   month                   4344 non-null   int64         
 3   month_name              4344 non-null   object        
 4   week_number             4344 non-null   int64         
 5   weekday_number          4344 non-null   int64         
 6   weekday_name            4344 non-null   object        
 7   is_weekend              4344 non-null   bool          
 8   is_us_federal_holiday   4344 non-null   bool          
 9   holiday_name            1008 non-null   object        
 10  is_religious_holiday    4344 non-null   bool          
 11  religious_holiday_name  888 non-null    object        
 12  religion                888 non-null    object  

## 4. Missing Values Analysis

#### Checking missing values in the calendar dataset

In [9]:
df_calendar.isnull().sum()

date                         0
year                         0
month                        0
month_name                   0
week_number                  0
weekday_number               0
weekday_name                 0
is_weekend                   0
is_us_federal_holiday        0
holiday_name              3336
is_religious_holiday         0
religious_holiday_name    3456
religion                  3456
religious_is_estimated       0
time                         0
dtype: int64

### Key Observations

- No missing values detected in core date fields (date, year, month, weekday)  
- Missing values in `holiday_name` and `religious_holiday_name` are expected  
  → they only apply when a holiday exists  

Overall, the dataset is structurally complete and consistent.

## 5. Duplicate Records Check

In [10]:
df_calendar.duplicated().sum()


np.int64(0)

No duplicate records found in the calendar dataset.

## 6. Date Coverage Validation

#### Reviewing date coverage and continuity

In [11]:
df_calendar["date"].min()

Timestamp('2023-01-01 00:00:00')

In [12]:
df_calendar["date"].max()

Timestamp('2023-06-30 00:00:00')

The dataset covers the period from January 1st, 2023 to June 30th, 2023.

## 7. Date Uniqueness Check

#### Validating uniqueness of calendar dates

In [13]:
df_calendar["date"].nunique(), df_calendar.shape[0]


(181, 4344)

The dataset contains 181 unique dates and 4344 total records.

This indicates that each date is represented multiple times, consistent with hourly granularity.

## 8. Hourly Granularity Validation

In [15]:
df_calendar['date'] = pd.to_datetime(df_calendar['date'])
df_calendar['hour_check'] = df_calendar['time'].apply(lambda t: t.hour)

# 1) Check: exactly 24 rows per day
rows_per_day = df_calendar.groupby(df_calendar['date'].dt.date).size()

print("Rows per day -> min:", rows_per_day.min(), "| max:", rows_per_day.max())
bad_rows_days = rows_per_day[rows_per_day != 24].sort_values()
bad_rows_days.head(30)

Rows per day -> min: 24 | max: 24


Series([], dtype: int64)

### Hourly Granularity Validation

Each day contains exactly 24 records (min = 24, max = 24).

No anomalies were detected in the number of records per day.

This confirms that the dataset has a consistent hourly granularity, which is essential for time-based analysis in the BI model.

## 8. Conclusion

The calendar dataset has been thoroughly validated and shows strong data quality.

- The dataset covers a continuous date range from January 1st, 2023 to June 30th, 2023  
- No missing values were found in core date-related fields  
- Missing values in holiday-related columns are expected and correspond to non-holiday periods  
- No duplicate records were detected  
- Each date is represented with exactly 24 records, confirming consistent hourly granularity  

Overall, the dataset is complete, consistent, and fully suitable for time-based analysis within the BI model.